s3/s3_transport/include/s3_transport.hpp

```diff
@@ -1067,58 +1067,99 @@ namespace irods::experimental::io::s3_transport
+
+                            if (part_number > number_of_parts) {
+                                break;
+                            }
 
-                        std::int64_t part_size = part_size_all_but_last_part;
+                            std::int64_t part_size = part_size_all_but_last_part;
 
-                        // give extra bytes to last part
-                        if (part_number == number_of_parts) {
-                            part_size += cache_file_size % number_of_parts;
-                        }
+                            // give extra bytes to last part
+                            if (part_number == number_of_parts) {
+                                part_size += (cache_file_size % number_of_parts);
+                            }
 
-                        off_t offset = (part_number - 1) * part_size_all_but_last_part;
+                            off_t offset = (part_number - 1) * part_size_all_but_last_part;
 
-                        irods::thread_pool::post(cache_flush_threads, [this, part_number, part_size, offset] () {
-                                    // upload part and read your part from cache file
-                                    s3_upload_part_worker_routine(true, part_number, part_size, offset);
-                        });
+                            irods::thread_pool::post(cache_flush_threads, [this, part_number, part_size, offset] () {
+                                        // upload part and read your part from cache file
+                                        s3_upload_part_worker_routine(true, part_number, part_size, offset);
+                            });
 
-                        ++part_number;
+                            ++part_number;
+                        }
+                        cache_flush_threads.join();
                     }
-                    cache_flush_threads.join();
+
+                    return_value = complete_multipart_upload();
+
+                } else {
+                    return_value = s3_upload_file(true);
                 }
 
-                return_value = complete_multipart_upload();
+                if (return_value == error_codes::UPLOAD_PART_TIMEOUT) {
+                    // reset saved errors
+                    shm_obj.atomic_exec([](auto& data) {
+                            data.last_error_code = error_codes::SUCCESS;
+                    });
+                    this->set_error(SUCCESS());
 
-            } else {
-                return_value = s3_upload_file(true);
+                    // halve the preferred_part_size and try again
+                    preferred_part_size >>= 1;
+
+                    logger::warn("{}:{} ({}) [[{}]] error in flushing cache file.  Trying a part size of {}.",
+                        __FILE__, __LINE__, __FUNCTION__, this->get_thread_identifier(), preferred_part_size);
+                } else {
+                    break_out = true;
```

Comment:
I think `break_out` may not be necessary if I'm reading this correctly.

This seems to be at the bottom of the `while (!break_out)` loop, so it could just break here. The only other place `break_out` is set is immediately before a `break` statement.

This is a really minor thing, so feel free to ignore.

Reply:
I went ahead and changed it to while(true) with breaks.  I made sure the comments before the while loop explain the exit criteria.